overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,260 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[string]*quota.Group) error {
+	// build the map of snap infos to options to provide to EnsureSnapServices
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+	for _, sn := range grp.Snaps {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return err
+		}
+
+		opts, err := SnapServiceOptions(st, sn, allGrps)
+		if err != nil {
+			return err
+		}
+
+		snapSvcMap[info] = opts
+	}
+
+	// TODO: the following lines should maybe be EnsureOptionsForDevice() or
+	// something since it is duplicated a few places
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: do we need to restart modified services ?
+	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+}
+
+func validateSnapForAddingToGroup(st *state.State, name string, allGrps map[string]*quota.Group) error {
+	// validate that the snap exists
+	_, err := snapstate.CurrentInfo(st, name)
+	if err != nil {
+		return err
+	}
+
+	// check that the snap is not already in a group
+	for _, grp := range allGrps {
+		if strutil.ListContains(grp.Snaps, name) {
+			return fmt.Errorf("snap already in quota group %q", grp.Name)
+		}
+	}
+	return nil
+}
+
+// CreateQuota attempts to create the specified quota group with the specified
+// snaps in it.
+// TODO: should this use something like QuotaGroupUpdate with fewer fields?
+func CreateQuota(st *state.State, name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
+	st.Lock()
+	defer st.Unlock()
+
+	// ensure that the quota group does not exist yet
```

Comment:
I think this comment should be moved down to line 102.

Reply:
done in https://github.com/snapcore/snapd/pull/10216